pkg/tfgen/docs.go

```diff
@@ -299,13 +296,66 @@ func getDocsForResource(g *Generator, source DocsSource, kind DocKind,
 				return doc, err
 			}
 
-			overlayArgsToArgs(sourceDocs, doc)
+			overlayArgsToArgs(sourceDocs, &doc)
 		}
 	}
 
 	return doc, nil
 }
 
+type expandedArguments map[string]*argumentNode
+type argumentNode struct {
+	docs     *argumentDocs
+	children expandedArguments
+}
+
+func (a *argumentNode) get(seg string) *argumentNode {
+	if a.children == nil {
+		a.children = expandedArguments{}
+	}
+	v, ok := a.children[seg]
+	if !ok {
+		v = &argumentNode{}
+		a.children[seg] = v
+	}
+	return v
+}
+
+func argumentTree(m map[docsPath]*argumentDocs) expandedArguments {
+	topLevel := &argumentNode{}
+	for k, v := range m {
+		a := topLevel
+		for _, segment := range k.parts() {
+			a = a.get(segment)
+		}
+		a.docs = v
+	}
+	return topLevel.children
+}
+
+func (t expandedArguments) collapse() map[docsPath]*argumentDocs {
+	m := make(map[docsPath]*argumentDocs, len(t))
+
+	var set func(docsPath, expandedArguments)
+	set = func(path docsPath, args expandedArguments) {
```

Comment:
`set` is declared and then used because the implementation of `set` calls `set`. By declaring before use, the body of `set` can refer to the `set` function, allowing recursion.

Reply:
Thanks; very slowly wrapping my head around go scopes.